### PR TITLE
Add a benchmark for luafun 

### PR DESCRIPTION
**Pull request by fsfod**
Added a benchmark for [luafun ](https://github.com/luafun/luafun) a functional programming library for Lua. The benchmark is the processing and stat collection of a JIT log collected from running the luacheck benchmark for 2 iterations.

### CODE REVIEW

File: benchmarks/luafun/bench.lua [hunk 1]
```diff
@@ -0,0 +1,136 @@
+local fun = require 'fun'
+local jitlog_capnp = require "jitlog_capnp"
+local bor = bit.bor
+local lshift = bit.lshift
+-- Enable table.new function
```
Comment (ltratt) [inline benchmarks/luafun/bench.lua]: I'm not sure this comment adds anything. Maybe remove?

File: benchmarks/luafun/bench.lua [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        if lastlua then
+            trace.stoploc = bor(lshift(lastlua.id, 16), trace.stoppc)
+        end
+    end
+
+end
+
+function run_iter(n)
+    for n=1, n do
+        init()
```
Comment (ltratt) [inline benchmarks/luafun/bench.lua]: Lots of blank lines after this. Are they all useful?

File: benchmarks/luafun/bench.lua [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+        local top_aborted = fun.iter(functions):max_by(function(f1, f2) return (f1.aborts > f2.aborts and f1) or f2 end)
+        stats.top_aborted = top_aborted
+        stats.top_aborted_list = fun.iter(aborted):filter(function(trace) return trace.lastlua == top_aborted end):totable()
+        assert(top_aborted.aborts == #stats.top_aborted_list)
+
+        stats.top_stitched = fun.iter(functions):max_by(function(f1, f2) return (f1.stitch_roots > f2.stitch_roots and f1) or f2 end)
+
+        --Build aborted trace stop locations counts
+        local aborted_locations = fun.iter(aborted):reduce(function(locs, trace)
```
Comment (ltratt) [inline benchmarks/luafun/bench.lua]: The indentation of this is really hard to read. At first I thought `end, {})` was a syntax error!

File: benchmarks/luafun/bench.lua [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            locs[trace.stoploc] = (locs[trace.stoploc] or 0) + 1
+            return locs
+        end, {})
+
+        stats.aborted_locations = aborted_locations
+        stats.most_aborted_location = fun.iter(aborted_locations):max_by(function(loc1, loc2)
+            return (aborted_locations[loc1] > aborted_locations[loc2] and loc1) or loc2
+        end)
```
Comment (ltratt) [inline benchmarks/luafun/bench.lua]: Ditto.

File: benchmarks/luafun/bench.lua [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+        --Build trace side exit start location counts
+        local sidexit_start_locations = fun.iter(completed):filter(function(trace) return trace.parentid ~= 0 end):reduce(function(locs, trace)
+            locs[trace.startloc] = (locs[trace.startloc] or 0) + 1
+            return locs
+        end, {})
```
Comment (ltratt) [inline benchmarks/luafun/bench.lua]: Ditto.

File: benchmarks/luafun/bench.lua (hunk 1 above)
Comment (fsfod) [reply benchmarks/luafun/bench.lua]: I just added it because the require result is not stored into anything it might not be obvious why.

Comment (ltratt) [reply benchmarks/luafun/bench.lua]: I assumed this was a standard Lua idiom?

Comment (fsfod) [reply benchmarks/luafun/bench.lua]: I've only seen LuaJIT do it for one or two functions

Comment (ltratt) [reply benchmarks/luafun/bench.lua]: OK.

File: benchmarks/luafun/bench.lua (hunk 3 above)
Comment (fsfod) [reply benchmarks/luafun/bench.lua]: The lines can get ridiculously long from call chains and trying to stuff multi-line function on a single line i think they would become unreadable. Even Roberto Ierusalimschy regrets not making a compact style for closures https://www.youtube.com/watch?v=uhBeH3r3FMQ&feature=youtu.be&t=132.

Comment (ltratt) [reply benchmarks/luafun/bench.lua]: I mean that we should do something like:

```lua
local x = function(y)
             blah
          end
```

to at least make the intent clear to readers.

File: benchmarks/luafun/bench.lua [hunk 6]
```diff
@@ -0,0 +1,178 @@
+local fun = require 'fun'
+local jitlog_capnp = require "jitlog_capnp"
+local bor = bit.bor
+local lshift = bit.lshift
+-- Enable table.new function
+require("table.new")
+
+local luafile = io.open("benchdata/jitlog.bin", "rb")
```
Comment (vext01) [inline benchmarks/luafun/bench.lua]: Is there anything like the Python context manager in lua? No worries if not.

```
with open("file.txt") as fh:
    stuff = fh.read()
# file was closed when the context manager inner scope expired
```

File: benchmarks/luafun/bench.lua [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        func.started_aborts = 0
+        func.stitch_roots = 0
+    end
+
+    -- Trace id remapping table
+    local idmap = {}
+    local tracecount = 1
+
+    for i, trace in ipairs(traces) do
+        --Build start and stop location keys for all the traces
```
Comment (vext01) [inline benchmarks/luafun/bench.lua]: Space after `--` to match other instances?

File: benchmarks/luafun/bench.lua [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        trace.startloc = bor(lshift(trace.startfunc, 16), trace.startpc)
+
+        local lastlua
+
```
Comment (vext01) [inline benchmarks/luafun/bench.lua]: Depending on taste, kill this blank line?

Comment (vext01) [reply benchmarks/luafun/bench.lua]: And generally review other whitespace throughout this file.

File: benchmarks/luafun/bench.lua [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+            return roots
+        end
+        local roots = fun.iter(completed):filter(function(trace) return trace.parentid == 0 end):reduce(collect_roots, {})
+        assert(next(roots))
+
+        for i, trace in ipairs(completed) do
+            local startfunc = functions[trace.startfunc]
+            startfunc.starts = startfunc.starts + 1
+
+            -- Map side traces to there root trace
```
Comment (vext01) [inline benchmarks/luafun/bench.lua]: "their"

File: benchmarks/luafun/bench.lua [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        end
+
+        local top_aborted = fun.iter(functions):max_by(function(f1, f2) return (f1.aborts > f2.aborts and f1) or f2 end)
+        stats.top_aborted = top_aborted
+        stats.top_aborted_list = fun.iter(aborted):filter(function(trace) return trace.lastlua == top_aborted end):totable()
+        assert(top_aborted.aborts == #stats.top_aborted_list)
+
+        stats.top_stitched = fun.iter(functions):max_by(function(f1, f2) return (f1.stitch_roots > f2.stitch_roots and f1) or f2 end)
+
+        --Build aborted trace stop locations stats
```
Comment (vext01) [inline benchmarks/luafun/bench.lua]: space after `--`

Comment (vext01) [reply benchmarks/luafun/bench.lua]: Please check all other instances.

**Comment by vext01** [comment]
Added a few shallow comments. I can't vouch for the correctness of the logic though.

File: benchmarks/luafun/bench.lua (hunk 10 above)
Comment (fsfod) [reply benchmarks/luafun/bench.lua]: fixed

**Comment by fsfod** [comment]
@vext01 and @ltratt can you take another look now.

**Comment by vext01** [comment]
I think there are several comments you've not yet responded to, no?

**Comment by fsfod** [comment]
Most them should be fixed but its subjective.

File: benchmarks/luafun/bench.lua (hunk 9 above)
Comment (fsfod) [reply benchmarks/luafun/bench.lua]: fixed

File: benchmarks/luafun/bench.lua (hunk 7 above)
Comment (fsfod) [reply benchmarks/luafun/bench.lua]: fixed

File: benchmarks/luafun/bench.lua (hunk 6 above)
Comment (vext01) [reply benchmarks/luafun/bench.lua]: So there is nothing like this for Lua?

File: benchmarks/luafun/bench.lua (hunk 9 above)
Comment (vext01) [reply benchmarks/luafun/bench.lua]: Did you push that fix? I don't see it.

> Map side traces to *their* root trace

Comment (fsfod) [reply benchmarks/luafun/bench.lua]: i was looking at the other one

File: benchmarks/luafun/bench.lua [hunk 11]
```diff
@@ -46,21 +45,20 @@ function init()
                 break
             end
         end
-
         trace.lastlua = lastlua
         if lastlua then
             trace.stoploc = bor(lshift(lastlua.id, 16), trace.stoppc)
         end
 
+        -- Map trace ids to a consistent increasing value based on there order in the log
```
Comment (ltratt) [inline benchmarks/luafun/bench.lua]: "there" -> "their"

File: benchmarks/luafun/bench.lua [hunk 12]
```diff
@@ -46,21 +45,20 @@ function init()
...
         end
-
         trace.lastlua = lastlua
         if lastlua then
             trace.stoploc = bor(lshift(lastlua.id, 16), trace.stoppc)
         end
 
+        -- Map trace ids to a consistent increasing value based on there order in the log
         local id = trace.id
+        -- Use the current seen successful trace count as the id for both aborted and successful traces
```
Comment (ltratt) [inline benchmarks/luafun/bench.lua]: I'm not sure how to parse "seen successful".

File: benchmarks/luafun/bench.lua (hunk 6 above)
Comment (fsfod) [reply benchmarks/luafun/bench.lua]: Nope only exception catching stuff like pcall and xpcall.

**Comment by ltratt** [comment]
OK, I made some comments in earlier commits which are fixed in later ones. Which is kinda cool in an odd way :)

File: benchmarks/luafun/bench.lua (hunk 12 above)
Comment (fsfod) [reply benchmarks/luafun/bench.lua]: The array the loop is going through is a mix of successful traces and aborted traces. tracecount is the current number of successful traces seen in the loop so far

Comment (fsfod) [reply benchmarks/luafun/bench.lua]: @ltratt what do you think it should say

Comment (ltratt) [reply benchmarks/luafun/bench.lua]: Something like "Use the current successful trace as count as..." maybe?

Comment (fsfod) [reply benchmarks/luafun/bench.lua]: what about "Use the current number of seen successful traces as the id for both aborted and successful traces" ?

Comment (ltratt) [reply benchmarks/luafun/bench.lua]: "Seen successful" just doesn't parse, to my ear at least -- you need to use a different phrase for that part.

Comment (fsfod) [reply benchmarks/luafun/bench.lua]: "Use the total number successful traces seen so far as the id for both aborted and successful traces" ?

Comment (ltratt) [reply benchmarks/luafun/bench.lua]: Works for me!

Comment (fsfod) [reply benchmarks/luafun/bench.lua]: fixed

File: benchmarks/luafun/bench.lua (hunk 6 above)
Comment (vext01) [reply benchmarks/luafun/bench.lua]: ok

**Comment by vext01** [comment]
I think that's my comments addressed. @ltratt if you are happy let's squash.

File: benchmarks/luafun/bench.lua [hunk 13]
```diff
@@ -52,7 +52,7 @@ function init()
 
         -- Map trace ids to a consistent increasing value based on their order in the log.
         local id = trace.id
-        -- Use the current seen successful trace count as the id for both aborted and successful traces.
+        -- Use the total number successful traces seen so far as the id for both aborted and successful traces.
```
Comment (ltratt) [inline benchmarks/luafun/bench.lua]: Sorry should be "total number of"

**Comment by ltratt** [comment]
@vext01 Once that little typo is fixed, please feel free to ask for a squash (in my absence).

**Comment by vext01** [comment]
ack

**Comment by fsfod** [comment]
@vext01 i fixed the comment

**Comment by vext01** [comment]
Cool. Go ahead and squash.

**Comment by fsfod** [comment]
@vext01  squashed

**Pull request merged**